core/src/main/scala/org/apache/spark/sql/delta/expressions/InterleaveBits.scala

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.expressions
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{ExpectsInputTypes, Expression}
+import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
+import org.apache.spark.sql.types.{BinaryType, DataType, IntegerType}
+
+
+/**
+ * Interleaves the bits of its input data in a round-robin fashion.
+ *
+ * If the input data is seen as a series of multidimensional points, this function computes the
+ * corresponding Z-values, in a way that's preserving data locality: input points that are close
+ * in the multidimensional space will be mapped to points that are close on the Z-order curve.
+ *
+ * The returned value is a byte array where the size of the array is 4 * num of input columns.
+ *
+ * @see https://en.wikipedia.org/wiki/Z-order_curve
+ *
+ * @note Only supports input expressions of type Int for now.
+ */
+case class InterleaveBits(children: Seq[Expression])
+  extends Expression with ExpectsInputTypes
+    with CodegenFallback /* TODO: implement doGenCode() */ {
+
+  private val n: Int = children.size
+
+  override def inputTypes: Seq[DataType] = Seq.fill(n)(IntegerType)
+
+  override def dataType: DataType = BinaryType
+
+  override def nullable: Boolean = false
+
+  /** Nulls in the input will be treated like this value */
+  val nullValue: Int = 0
+
+  private val childrenArray: Array[Expression] = children.toArray
+
+  private val ints = new Array[Int](n)
+
+  override def eval(input: InternalRow): Any = {
+    var i = 0
+    while (i < n) {
+      val int = childrenArray(i).eval(input) match {
+        case null => nullValue
+        case int: Int => int
+        case any => throw new IllegalArgumentException(
+          s"${this.getClass.getSimpleName} expects only inputs of type Int, but got: " +
+            s"$any of type${any.getClass.getSimpleName}")
+      }
+      ints.update(i, int)
+      i += 1
+    }
+
+    val ret = new Array[Byte](n * 4)
```

Comment:
The Apache [Kyuubi](https://github.com/apache/incubator-kyuubi/blob/master/extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/zorder/ZorderBytesUtils.scala) community has implemented a fast approach based on this [documentation](http://graphics.stanford.edu/~seander/bithacks.html#InterleaveTableObvious).


I also ran some benchmarks based on my forked [repository](https://github.com/delta-io/delta/compare/master...cfmcgrady:fast-interleave-bits?expand=1).


UT passed.

```
[info] MultiDimClusteringFunctionsSuite:
[info] - range_partition_id(): simple (7 seconds, 152 milliseconds)
[info] - range_partition_id(): two columns (2 seconds, 231 milliseconds)
[info] - range_partition_id(): corner cases (1 second, 209 milliseconds)
[info] - interleave_bits(): 1 input = cast to binary (137 milliseconds)
[info] - interleave_bits(): arbitrary num inputs (343 milliseconds)
[info] - interleave_bits(): corner cases (313 milliseconds)
[info] - interleave_bits(range_partition_ids) (645 milliseconds)
[info] Run completed in 18 seconds, 470 milliseconds.
[info] Total number of tests run: 7
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 7, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

```
[info] InterleaveBitsSuite:
[info] - 0 inputs (2 seconds, 421 milliseconds)
[info] - 1 input (339 milliseconds)
[info] - 2 inputs (26 milliseconds)
[info] - 3 inputs (24 milliseconds)
[info] - nulls (180 milliseconds)
[info] - consistency (358 milliseconds)
[info] - supported types (2 milliseconds)
[info] Run completed in 5 seconds, 349 milliseconds.
[info] Total number of tests run: 8
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 7, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

```scala
object InterleaveBitsBenchmark extends BenchmarkBase {

  private val numRows = 1 * 1000 * 1000

  private def randomInt(numColumns: Int): Seq[Array[Int]] = {
    (1 to numRows).map { l =>
      val arr = new Array[Int](numColumns)
      (0 until numColumns).foreach(col => arr(col) = l)
      arr
    }
  }

  def createExpression(numColumns: Int): Expression = {

    val inputs = (0 until numColumns).map { i =>
      $"c_$i".int.at(i)
    }
    InterleaveBits(inputs)
  }

  protected def create_row(values: Any*): InternalRow = {
    InternalRow.fromSeq(values.map(CatalystTypeConverters.convertToCatalyst))
  }

  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
    val benchmark =
      new Benchmark(s"$numRows rows interleave bits benchmark", numRows, output = output)
    benchmark.addCase("1 int columns benchmark", 3) { _ =>
      val interleaveBits = createExpression(1)
      randomInt(1).foreach { input =>
        interleaveBits.eval(create_row(input: _*))
      }
    }

    benchmark.addCase("2 int columns benchmark", 3) { _ =>
      val interleaveBits = createExpression(2)
      randomInt(2).foreach { input =>
        interleaveBits.eval(create_row(input: _*))
      }
    }

    benchmark.addCase("3 int columns benchmark", 3) { _ =>
      val interleaveBits = createExpression(3)
      randomInt(3).foreach { input =>
        interleaveBits.eval(create_row(input: _*))
      }
    }

    benchmark.addCase("4 int columns benchmark", 3) { _ =>
      val interleaveBits = createExpression(4)
      randomInt(4).foreach { input =>
        interleaveBits.eval(create_row(input: _*))
      }
    }
    benchmark.run()
  }
}
```


- Delta current implementation


```
1000000 rows interleave bits benchmark:   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
1 int columns benchmark                             565            593          33          1.8         565.1       1.0X
2 int columns benchmark                             816            823           8          1.2         816.4       0.7X
3 int columns benchmark                            1054           1056           2          0.9        1054.1       0.5X
4 int columns benchmark                            1275           1289          23          0.8        1274.5       0.4X
```

- Fast Approach

```
1000000 rows interleave bits benchmark:   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
1 int columns benchmark                             434            445          15          2.3         434.3       1.0X
2 int columns benchmark                             707            819         154          1.4         706.6       0.6X
3 int columns benchmark                             777            783           6          1.3         776.8       0.6X
4 int columns benchmark                             954            961           8          1.0         954.1       0.5X
```

Reply:
Hi @cfmcgrady, this is awesome. Can you post the new implementation as a PR?